lib/vquic/quiche.c

```diff
@@ -623,11 +634,16 @@ static CURLcode http_request(struct connectdata *conn, const void *mem,
   for(i = 1; i < len; ++i) {
     if(hdbuf[i] == '\n' && hdbuf[i - 1] == '\r') {
       ++nheader;
+      if(i >= 3 && hdbuf[i - 2] == '\n' && hdbuf[i - 3] == '\r') {
+        break;
```

Comment:
This seems unrelated to this bugfix, right? Why is this change necessary?

Reply:
@bagder  It's related. I just found if I upload small file, the request body would follow the http header, with two `\r\n` splited, just like the http protocol. In this case, if you don't stop when you meet two continuous `\r\n`, and by coincidence there's another `\r\n` in your request body, then you could get the wrong header count.

Here's my gdb print:

```
$ LD_LIBRARY_PATH=$PWD/lib/.libs/ rust-gdb --args ./src/.libs/curl --tlsv1.3 --http3 -vk -X POST --data-binary "@1k-file" https://blog.cloudflare.com/
GNU gdb (Ubuntu 7.7.1-0ubuntu5~14.04.3) 7.7.1
Copyright (C) 2014 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./src/.libs/curl...done.
(gdb) break quiche.c:633
No source file named quiche.c.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (quiche.c:633) pending.
(gdb) run
Starting program: /root/curl/src/.libs/curl --tlsv1.3 --http3 -vk -X POST --data-binary @1k-file https://blog.cloudflare.com/
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Note: Unnecessary use of -X or --request, POST is already inferred.
* STATE: INIT => CONNECT handle 0x64b548; line 1654 (connection #-5000)
* Added connection 0. The cache now contains 1 members
[New Thread 0x7ffff6691700 (LWP 1215)]
* STATE: CONNECT => WAITRESOLVE handle 0x64b548; line 1700 (connection #0)
[Thread 0x7ffff6691700 (LWP 1215) exited]
* family0 == v4, family1 == v6
*   Trying 104.18.26.46:443...
* Sent QUIC client Initial, ALPN: h3-29,h3-28,h3-27
* STATE: WAITRESOLVE => WAITCONNECT handle 0x64b548; line 1782 (connection #0)
* quiche established connection!
* Connected to blog.cloudflare.com (104.18.26.46) port 443 (#0)
* Marked for [keep alive]: HTTP/3 default
* STATE: WAITCONNECT => SENDPROTOCONNECT handle 0x64b548; line 1848 (connection #0)
* STATE: SENDPROTOCONNECT => DO handle 0x64b548; line 1869 (connection #0)

Breakpoint 1, http_request (conn=0x64a5a8, mem=0x6794f8, len=1181, written=0x7fffffffdcc8) at vquic/quiche.c:633
633       nheader = 0;
(gdb) p hdbuf
$1 = 0x6794f8 "POST / HTTP/3\r\nHost: blog.cloudflare.com\r\nUser-Agent: curl/7.74.0-DEV\r\nAccept: */*\r\nContent-Length: 1024\r\nContent-Type: application/x-www-form-urlencoded\r\n\r\nt8\215\017\223t\352\201\314\323-ZQv\226\336\304\332\356ZV&\323n\330e\341\364V\357\372\317X\324q?\234\272\252\243*+\024"...
```